### PR TITLE
Add emptyElementParsePolicy to main DFDL namespace.

### PR DESCRIPTION
**Pull request by mike-mcgann**
The emptyElementParsePolicy is currently implemented as a property that is an extension to DFDL. As of DFDL 1.0, this property is now available in the main schema. This property has now been added and a deprecation message is emitted when the old property is used. The property value of `treatAsMissing` has been renamed to `treatAsAbsent` in DFDL 1.0.

[DAFFODIL-2496](https://issues.apache.org/jira/projects/DAFFODIL/issues/DAFFODIL-2496)

### CODE REVIEW

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/SequenceChild.scala
Comment (stevedlawrence) [inline daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/SequenceChild.scala:441]: Should we remove the `x` in the `dfdlx` prefix here since the extension is deprecated?

File: daffodil-propgen/src/main/resources/org/apache/daffodil/xsd/dfdlx.xsd
Comment (stevedlawrence) [inline daffodil-propgen/src/main/resources/org/apache/daffodil/xsd/dfdlx.xsd:102]: I don't think `treatAsAbsent` should be added here? This simple type restriction is used only for the dfdx extension property, which uses `treatAsMissing` instead of `treatAsAbsent`. If someone wants to use `treatAsAbsent`, let's force them to use the non-extension property.

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceChildBases.scala
Comment (stevedlawrence) [inline daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceChildBases.scala:81]: `dfdlx` should be `dfdl` here.

File: daffodil-lib/src/main/scala/org/apache/daffodil/schema/annotation/props/ByHandMixins.scala
Comment (stevedlawrence) [inline daffodil-lib/src/main/scala/org/apache/daffodil/schema/annotation/props/ByHandMixins.scala:439]: Around line 470 in this file, we have this comment:
```scala
// This property is an extension, so we don't want to require users to
// add this property just to silence this warning, especially since the
// property might change in the future. So silently use a default without
// outputting a warning. We may want to turn this warning on when the
// property makes its way into the official DFDL spec.
//SDW(
//  WarnID.EmptyElementParsePolicyError,
//  "Property 'dfdlx:emptyElementParsePolicy' is required but not defined, using tunable '%s' by default.",
//  defaultEmptyElementParsePolicy)
```
Do we want to re-enable this warning on now that this property is in the spec?

Do we also want to remove the defaulting behavior? I imagine no, since it might break a number of schemas? I guess after having this warning on for some number of releases we can flip the switch and require this property.

Do we want to add a value to DFDLGeneralFormat.dfdl.xsd? Is there an obvious choice or is it really dependent on the file format?

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml
Comment (stevedlawrence) [inline daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml:197]: Does IBM support this property now? Do we no longer need to remove this property for cross tests to work?

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml
Comment (stevedlawrence) [inline daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml:281]: Same question here, does this still need to be removed for IBM?

**Review by stevedlawrence — COMMENTED** [review]
Looks good! Just a couple minor comments and questions

File: daffodil-propgen/src/main/resources/org/apache/daffodil/xsd/dfdlx.xsd
Comment (mike-mcgann) [reply daffodil-propgen/src/main/resources/org/apache/daffodil/xsd/dfdlx.xsd:102]: This is a bit of a hack and I couldn't find a way to easily make this cleaner. Let me know if you can think of a better way. 

The issue is that we want to support the new and old attribute, for now, at the same time. An element could have a dfdl:emptyElementParsePolicy or a dfdlx:emptyElementParsePolicy but the property generator fabricates variables using the unprefixed name. Therefore we cannot easily have both of these properties with different types since they map to the same symbol. The schema validator correctly uses the empty/absent enumeration when evaluating the attribute with the dfdl prefix but then the parser uses the dfdlx enumeration definition (which is manually constructed and ignored by the property generator). The absent value was added to the dfdlx enumeration to support both new and old since there can be only one type. I guess the cleanest way would be to change the property generator to use fully qualified attribute names, or at least use them when necessary, but I'm not sure what the wider consequences of that change would be.

File: daffodil-lib/src/main/scala/org/apache/daffodil/schema/annotation/props/ByHandMixins.scala
Comment (stevedlawrence) [reply daffodil-lib/src/main/scala/org/apache/daffodil/schema/annotation/props/ByHandMixins.scala:439]: @mbeckerle thoughts on the above questions?

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml
Comment (mike-mcgann) [reply daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml:197]: I refactored this a bit.

Schema 4 now uses the old dfdlx:emptyElementParsePolicy and the IBM test for that is the same as before. I added a new test for daffodil only that checks that a parse error and a deprecation warning is issued. 

Schema 5 uses the new dfdl:emptyElementParsePolicy and that test only runs on daffodil. I'm not sure what happens on IBM for that.

File: daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/SequenceChild.scala
Comment (mike-mcgann) [reply daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/SequenceChild.scala]: Updated

File: daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceChildBases.scala
Comment (mike-mcgann) [reply daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceChildBases.scala]: Updated

File: daffodil-propgen/src/main/resources/org/apache/daffodil/xsd/dfdlx.xsd
Comment (stevedlawrence) [reply daffodil-propgen/src/main/resources/org/apache/daffodil/xsd/dfdlx.xsd:102]: I think I'm still confused.

I understand why ByHandMixin.scala needs all three enumerations since both dfdl and dfdlx properties use the same EmptyElementParsePolicy object to convert a string to the type.

> but the property generator fabricates variables using the unprefixed name

The property generator excludes the EmptyElementParsePolicy type, so doesn't it completely ignore the definitions in this dfdlx.xsd file? So aren't the enumerations in this file just about doing DFDL schema validation and nothing else? And since treatAsAbsent isn't valid in he dfdlx property shouldn't it be removed here and only be defined in the DFDL_partX xsd file?

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml
Comment (mbeckerle) [reply daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml:197]: The property dfdl:emptyElementParsePolicy value 'treatAsEmpty' is an optional feature of DFDL.
IBM does not support 'treatAsEmpty' behavior, only 'treatAsAbsent'.

As for whether they added this property, let's test it. We have the IBM DFDL Cross tester.  https://github.com/OpenDFDL/ibmDFDLCrossTester

File: daffodil-propgen/src/main/resources/org/apache/daffodil/xsd/dfdlx.xsd
Comment (mike-mcgann) [reply daffodil-propgen/src/main/resources/org/apache/daffodil/xsd/dfdlx.xsd]: I think I was the one that was confused. I didn't realize that with the exclusion that the xsd file was completely ignored. I've removed that value from the schema and all the tests run fine. Thanks for that clarification.

File: daffodil-lib/src/main/scala/org/apache/daffodil/schema/annotation/props/ByHandMixins.scala
Comment (mbeckerle) [reply daffodil-lib/src/main/scala/org/apache/daffodil/schema/annotation/props/ByHandMixins.scala:439]: We should re-enable this warning, yes. Making the property required right now would break people's schemas. 

We do need to also add this property to DFDLGeneralFormat,  with value treatAsEmpty. But this may need to be in Daffodil-only part of that if IBM DFDL doesn't tolerate the property.

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml
Comment (tuxji) [reply daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml:197]: Unfortunately, https://www.ibm.com/docs/en/integration-bus/10.0?topic=dfdl-unsupported-features does not say anything about emptyElementParsePolicy.

**Review by tuxji — CHANGES_REQUESTED** [review]
I think this PR needs to re-enable a warning, and I hope it won't break any tests, but we need to check first.

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/MultipartBody.dfdl.xsd
Comment (mbeckerle) [inline daffodil-test/src/test/resources/org/apache/daffodil/usertests/MultipartBody.dfdl.xsd:57]: Indentation in this file is totally goofed up. Rather than fix it, and have whitespace fixing be part of this PR, let's accumulate whitespace fixups on a ticket. 

I already added this file to this ticket: https://issues.apache.org/jira/browse/DAFFODIL-2759

File: daffodil-lib/src/main/scala/org/apache/daffodil/schema/annotation/props/ByHandMixins.scala
Comment (stevedlawrence) [reply daffodil-lib/src/main/scala/org/apache/daffodil/schema/annotation/props/ByHandMixins.scala:439]: > we want to add the property to DFDLGeneralFormat.dfdl.xsd and give it the same 'treatAsEmpty' default value. 

I just thought of one potential issue when adding to DFDLGeneralFormat. If IBM DFDL doesn't support this new property then any schemas that used DFDLGeneralFormat.dfdl.xsd might now error when cross run with IBM? To ensure backwards compat, maybe we don't want to add the property to the file, especially since it's optional (as Mike says) and we already have built in defaulting behavior? Though as you point out, that would mean lots of schemas would need to manually add this property to suppress the warning.

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml
Comment (mbeckerle) [inline daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml:281]: Should this be just no-prefix emptyElementParsePolicy='treatAsAbsent' now?

I probably wrote this test originally, but I don't understand the comment "remove extension property for IBM cross tests" since this is used by an IBM cross test below. It must be that IBM DFDL is able to ignore the dfdlx extensions or this wouldn't have worked in test test_sep_ssp_never_4_ibm.

So I think the comment line "remove extension property for IBM cross tests" is spurious now.

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml
Comment (mbeckerle) [inline daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml:284]: change comment to 'treatAsAbsent'.

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml
Comment (mbeckerle) [inline daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml:299]: change comment to treatAsAbsent, and remove parenthetic note about fixing 2496.

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml
Comment (mbeckerle) [inline daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml:331]: I hate it when github won't let me put a comment on the line I want to. They have, in their infinite wisdom, descided that if you didn't change a line or the line isn't close enough to a change, no comment is allowed. 

Below a tdml:error is expecting 'treatAsMissing'. I believe this should change, per the comment, now.

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml
Comment (mbeckerle) [inline daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml:412]: While you are at it, add a final new-line.

**Review by mbeckerle — CHANGES_REQUESTED** [review]
Small clean ups I think, and we need to find out whether IBM DFDL accepts the real property name or not yet.

File: daffodil-lib/src/main/scala/org/apache/daffodil/schema/annotation/props/ByHandMixins.scala
Comment (tuxji) [reply daffodil-lib/src/main/scala/org/apache/daffodil/schema/annotation/props/ByHandMixins.scala:439]: IBM should have its own DFDLGeneralFormat.dfdl.xsd so that schemas can import DFDLGeneralFormat.dfdl.xsd without worrying about problems like that, yes?  If the IBM cross tester is using Daffodil's DFDLGeneralFormat.dfdl.xsd with IBM's DFDL processor, that seems like a behavior which needs to be changed, don't you think?

Comment (stevedlawrence) [reply daffodil-lib/src/main/scala/org/apache/daffodil/schema/annotation/props/ByHandMixins.scala:439]: DFDLGeneralFormat.dfdl.xsd is a thing Daffodil invented, I don't think IBM has any concept of it. DFDL schemas use the full path of `org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd `, which is daffodil specific and what the cross tester will use since that is what schemas import.

I guess we do have two different defineFormats: `DFDLGeneralFormat` and `DFDLGeneralFormatPortable`. We could put the new property in the DFDLGeneralFormat define, but many schemas (including most of our tests) don't use the "portable" variant and will probably fail if cross tested. Maybe that's okay?

Comment (mbeckerle) [reply daffodil-lib/src/main/scala/org/apache/daffodil/schema/annotation/props/ByHandMixins.scala:439]: Yes, tests that are portable have to use GeneralFormatPortable because there are some other subtle incompatibilities between IBM DFDL and Daffodil - things about time zones (like Z vs. -00:00 suffixes equivalent values, but not identical, so the same test doesn't compare correctly), lack of support for encodingErrorPolicy 'replace', etc.

Comment (mbeckerle) [reply daffodil-lib/src/main/scala/org/apache/daffodil/schema/annotation/props/ByHandMixins.scala:439]: To clarify, both GeneralFormat and GeneralFormatPortable are defined in DFDLGeneralFormat.dfdl.xsd. They share a base format with all the fully portable properties.

Comment (mike-mcgann) [reply daffodil-lib/src/main/scala/org/apache/daffodil/schema/annotation/props/ByHandMixins.scala:439]: I've added the property to `DFDLGeneralFormat.dfdl.xsd` and re-enabled the warning. This required an update to a test where it was relying on the property to be missing and defaulting. I've updated that test to use a different schema instead of using the tuneable.

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/MultipartBody.dfdl.xsd
Comment (mike-mcgann) [reply daffodil-test/src/test/resources/org/apache/daffodil/usertests/MultipartBody.dfdl.xsd]: Updated.

Comment (mike-mcgann) [reply daffodil-test/src/test/resources/org/apache/daffodil/usertests/MultipartBody.dfdl.xsd]: I went ahead and updated this from an earlier request.

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml
Comment (mike-mcgann) [reply daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml]: Added

**Review by tuxji — APPROVED** [review]
+1

Have you run the IBM cross tester successfully?  If so, changes look good now, but wait for Steve or Mike to review too.

**Review by stevedlawrence — APPROVED** [review]
+1 :+1: 

> Have you run the IBM cross tester successfully?

Yeah, it would be good to confirm that a schema using GeneralFormatPortable still works when used with IBM DFDL.

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml
Comment (mike-mcgann) [reply daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml]: @mbeckerle helped me out this morning getting the IBM test runner working and we found out that this property is not supported at this time. We updated the tests to reflect this.

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml
Comment (mike-mcgann) [reply daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml]: Updated during our meeting.

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml
Comment (mike-mcgann) [reply daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml]: Updated during our meeting

File: daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml
Comment (mike-mcgann) [reply daffodil-test/src/test/resources/org/apache/daffodil/usertests/SepTests.tdml]: This is either updated or got revised out

File: daffodil-lib/src/main/resources/org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd
Comment (stevedlawrence) [inline daffodil-lib/src/main/resources/org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd:107]: I see you moved the property from GeneralFormatOriginal to GeneralFormat in this update. I guess this is so that the property is not defined in GeneralFormatPortable, so as long as schemas use the portabale format, IBM DFDL works as expected?

Comment (mike-mcgann) [reply daffodil-lib/src/main/resources/org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd:107]: Yes.

**Review by mbeckerle — APPROVED** [review]
+1

**Comment by tuxji** [comment]
I just noticed that this PR needs to be merged by a Daffodil committer.  I confirm that all PR comments have been resolved, @mike-mcgann has squashed the PR to one commit, and all tests have passed successfully.

**Pull request merged**
